### PR TITLE
Exposing Helpful Anomaly Detection Metadata from Anomaly Strategies (ie Anomaly Check Range/Thresholds) through backwards compatible function

### PR DESCRIPTION
**Pull request by arsenalgunnershubert777**
*Description of changes:*

This PR adds functionality to expose anomaly detection metadata for anomaly checks.

There is a legacy PR [here](https://github.com/awslabs/deequ/pull/525), and this PR addresses feedback to make the anomaly detection metadata functionality backwards compatible.

- [AnomalyDetectionExtendedResult](https://github.com/arsenalgunnershubert777/deequ/blob/5da25c4d3738a5ac32f6cb2f1edeb9262bec6b2a/src/main/scala/com/amazon/deequ/anomalydetection/ExtendedDetectionResult.scala#L134) is a newly exposed field that contains metadata for anomaly check runs. The user is able to see these details from ConstraintResult.
- The user can see those extended results when using a new [addAnomalyCheckWithExtendedResults](https://github.com/arsenalgunnershubert777/deequ/blob/5da25c4d3738a5ac32f6cb2f1edeb9262bec6b2a/src/main/scala/com/amazon/deequ/VerificationRunBuilder.scala#L244-L261) method in the verification suite builder (instead of [addAnomalyCheck](https://github.com/arsenalgunnershubert777/deequ/blob/5da25c4d3738a5ac32f6cb2f1edeb9262bec6b2a/src/main/scala/com/amazon/deequ/VerificationRunBuilder.scala#L226-L242))
- The original [addAnomalyCheck](https://github.com/arsenalgunnershubert777/deequ/blob/5da25c4d3738a5ac32f6cb2f1edeb9262bec6b2a/src/main/scala/com/amazon/deequ/VerificationRunBuilder.scala#L226-L242) functionality works the same (though it now internally uses the new detectWithExtendedResults function see point about that below)
- There are the following changes to make this happen
  - First of all a new [AnomalyDetectionDataPoint class](https://github.com/arsenalgunnershubert777/deequ/blob/5da25c4d3738a5ac32f6cb2f1edeb9262bec6b2a/src/main/scala/com/amazon/deequ/anomalydetection/ExtendedDetectionResult.scala#L42-L48)
    - This class has the following fields:
      - dataMetricValue: The metric value that is the data point.
      - anomalyMetricValue: The metric value that is being checked for the anomaly detection strategy, which isn't always equal to the dataMetricValue. 
      - anomalyCheckRange: The range of bounds used in the anomaly check, the anomalyMetricValue is compared to this range. 
      - isAnomaly: If the anomalyMetricValue is outside the anomalyCheckRange, this is true.
      - confidence: The confidence of the anomaly detection.
      - detail: An optional detail message.
    - [AnomalyDetectionExtendedResult](https://github.com/arsenalgunnershubert777/deequ/blob/5da25c4d3738a5ac32f6cb2f1edeb9262bec6b2a/src/main/scala/com/amazon/deequ/anomalydetection/ExtendedDetectionResult.scala#L134) is currently just a wrapper for this [AnomalyDetectionDataPoint](https://github.com/arsenalgunnershubert777/deequ/blob/5da25c4d3738a5ac32f6cb2f1edeb9262bec6b2a/src/main/scala/com/amazon/deequ/anomalydetection/ExtendedDetectionResult.scala#L42-L48) class.
  - All the anomaly detection strategies extend a new trait called [AnomalyDetectionStrategyWithExtendedResults](https://github.com/arsenalgunnershubert777/deequ/blob/5da25c4d3738a5ac32f6cb2f1edeb9262bec6b2a/src/main/scala/com/amazon/deequ/anomalydetection/AnomalyDetectionStrategy.scala#L33-L46)
    - In the batch normal strategy for example, the new [detectWithExtendedResults function](https://github.com/arsenalgunnershubert777/deequ/blob/5da25c4d3738a5ac32f6cb2f1edeb9262bec6b2a/src/main/scala/com/amazon/deequ/anomalydetection/BatchNormalStrategy.scala#L74-L119) extends the new trait.
  - The new [detectWithExtendedResults function](https://github.com/arsenalgunnershubert777/deequ/blob/5da25c4d3738a5ac32f6cb2f1edeb9262bec6b2a/src/main/scala/com/amazon/deequ/anomalydetection/BatchNormalStrategy.scala#L74-L119) is also updated to be the function that the regular [detect function](https://github.com/arsenalgunnershubert777/deequ/blob/5da25c4d3738a5ac32f6cb2f1edeb9262bec6b2a/src/main/scala/com/amazon/deequ/anomalydetection/BatchNormalStrategy.scala#L55-L64) uses after mapping from the AnomalyDetectionDataPoint to the original Anomaly class, since they are using the same underlying calculations. If preferred to keep them independent, I can separate those. 
  - Using the detectWithExtendedResults function [all the Anomaly Strategies](https://github.com/arsenalgunnershubert777/deequ/tree/5da25c4d3738a5ac32f6cb2f1edeb9262bec6b2a/src/main/scala/com/amazon/deequ/anomalydetection) output all the AnomalyDetectionDataPoints with those extra Anomaly details corresponding to all the data points in the search interval via [ExtendedDetectionResults](https://github.com/arsenalgunnershubert777/deequ/blob/5da25c4d3738a5ac32f6cb2f1edeb9262bec6b2a/src/main/scala/com/amazon/deequ/anomalydetection/ExtendedDetectionResult.scala#L120)
  - The last AnomalyDetectionDataPoint which corresponds to the new metric being checked is passed along the through the following:  
    - The new [isNewestPointNonAnomalousWithExtendedResults function](https://github.com/arsenalgunnershubert777/deequ/blob/5da25c4d3738a5ac32f6cb2f1edeb9262bec6b2a/src/main/scala/com/amazon/deequ/checks/Check.scala#L609-L629)
    - The new [Check.isNewestPointNonAnomalousWithExtendedResults function](https://github.com/arsenalgunnershubert777/deequ/blob/5da25c4d3738a5ac32f6cb2f1edeb9262bec6b2a/src/main/scala/com/amazon/deequ/checks/Check.scala#L1425-L1487) called by the above
    - The new [getNewestPointAnomalyResults](https://github.com/arsenalgunnershubert777/deequ/blob/5da25c4d3738a5ac32f6cb2f1edeb9262bec6b2a/src/main/scala/com/amazon/deequ/checks/Check.scala#L1496) function called by the above which [returns](https://github.com/arsenalgunnershubert777/deequ/blob/5da25c4d3738a5ac32f6cb2f1edeb9262bec6b2a/src/main/scala/com/amazon/deequ/checks/Check.scala#L619-L626) a new kind of Anomaly assertion function
    - This new anomaly assertion function called by the above returns a new [AnomalyDetectionAssertionResult](https://github.com/arsenalgunnershubert777/deequ/blob/5da25c4d3738a5ac32f6cb2f1edeb9262bec6b2a/src/main/scala/com/amazon/deequ/anomalydetection/ExtendedDetectionResult.scala#L142). 
    - The new AnomalyDetectionAssertionResult [contains both an assertion boolean as well as the AnomalyDetectionExtendedResult wrapping the data point](https://github.com/arsenalgunnershubert777/deequ/blob/5da25c4d3738a5ac32f6cb2f1edeb9262bec6b2a/src/main/scala/com/amazon/deequ/checks/Check.scala#L1512-L1514) of the newest data point. The assertion boolean is determined from that data point's isAnomaly field.
      - This is different from the previous anomaly assertion function [which just returns a boolean](https://github.com/arsenalgunnershubert777/deequ/blob/5da25c4d3738a5ac32f6cb2f1edeb9262bec6b2a/src/main/scala/com/amazon/deequ/checks/Check.scala#L1400), because this allows the passing through of the anomaly detection data point.
    - This new anomaly assertion function uses a new [AnomalyExtendedResultsConstraint class](https://github.com/arsenalgunnershubert777/deequ/blob/5da25c4d3738a5ac32f6cb2f1edeb9262bec6b2a/src/main/scala/com/amazon/deequ/constraints/AnomalyExtendedResultsConstraint.scala) to [handle the assertion function](https://github.com/arsenalgunnershubert777/deequ/blob/5da25c4d3738a5ac32f6cb2f1edeb9262bec6b2a/src/main/scala/com/amazon/deequ/constraints/AnomalyExtendedResultsConstraint.scala#L81-L91). 
    - In the future if the Anomaly Check focuses on multiple datapoints instead of just one, the [AnomalyDetectionExtendedResult](https://github.com/arsenalgunnershubert777/deequ/blob/5da25c4d3738a5ac32f6cb2f1edeb9262bec6b2a/src/main/scala/com/amazon/deequ/anomalydetection/ExtendedDetectionResult.scala#L134) can add another field and be adjusted accordingly.
  - The anomaly detection metadata is [added to the ConstraintResult class](https://github.com/arsenalgunnershubert777/deequ/blob/5da25c4d3738a5ac32f6cb2f1edeb9262bec6b2a/src/main/scala/com/amazon/deequ/constraints/Constraint.scala#L50) as an optional field
    - Maybe in the future this can be separated into a different class that inherits from a common trait since the AnomalyDetectionMetadata does not pertain to other non anomaly checks, would take a lot more refactoring/changes
  - Tests are updated to reflect the above changes
  - Documentation with links to an example executable is added in this [anomaly detection with extended results page](https://github.com/arsenalgunnershubert777/deequ/blob/5da25c4d3738a5ac32f6cb2f1edeb9262bec6b2a/src/main/scala/com/amazon/deequ/examples/anomaly_detection_with_extended_results_example.md) that is linked from the previous [anomaly detection page](https://github.com/arsenalgunnershubert777/deequ/blob/5da25c4d3738a5ac32f6cb2f1edeb9262bec6b2a/src/main/scala/com/amazon/deequ/examples/anomaly_detection_example.md)

Let me know if anyone has any questions/feedback, thanks!

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by rdsharma26** [comment]
@arsenalgunnershubert777 Thanks for the PR! Please allow for additional time to review the PR given the amount of changes. In the meantime, would it be possible to rebase your branch off of latest `master` so that it reduces the number of commits in the PR?

**Comment by arsenalgunnershubert777** [comment]
Hey @rdsharma26 thanks for reaching out. I think I have some issues with the conflicts (and I also can improve my understanding of Git)
So for now I may just create a clean branch and apply the changes on top of that manually and open a new PR, so there's just 1 commit

**Comment by rdsharma26** [comment]
Thanks @arsenalgunnershubert777 
After syncing your forked repository with the main repository, you could try:

```
git checkout master
git pull
git checkout exposeAnomalyThresholdBackwardsCompatible
git rebase master
git push origin exposeAnomalyThresholdBackwardsCompatible
``

**Comment by arsenalgunnershubert777** [comment]
@rdsharma26 got it yea when I run that it has more conflicts, and I notice some classes have reverted to previous states

**Comment by rdsharma26** [comment]
@arsenalgunnershubert777 We will review this PR this week. Thank you for the patience.

File: src/main/scala/com/amazon/deequ/checks/Check.scala [hunk 1]
```diff
@@ -1355,4 +1399,117 @@ object Check {
...
+   * @param beforeDate               The maximum dateTime of previous AnalysisResults to use for
+   *                                 the Anomaly Detection.
+   * @param afterDate                The minimum dateTime of previous AnalysisResults to use for
+   *                                 the Anomaly Detection.
+   * @param currentMetricValue       current metric value.
+   * @return The AnomalyDetectionAssertionResult with the boolean if the newest data point is anomalous
+   *         along with the AnomalyDetectionExtendedResult object which contains the
+   *         anomaly detection extended result details.
+   */
+  private[deequ] def isNewestPointNonAnomalousWithExtendedResults[S <: State[S]](
```
Comment (rdsharma26) [inline src/main/scala/com/amazon/deequ/checks/Check.scala:1425]: Should we change the name of this method? The `is<>` implies returning a boolean. Since we are returning a case class, perhaps we can update the method name to imply that.

**Review by rdsharma26 — COMMENTED** [review]
@arsenalgunnershubert777

Thanks again for this PR. It is detailed and contains extensive documentation and testing.

I've taken a first pass at the changes and left some comments. I see opportunities to reduce repetitive code by reducing the number of new case classes that are being introduced. The new APIs being introduced look good. Can you have a look at my feedback and send another revision? Thank you!

**Comment by arsenalgunnershubert777** [comment]
@rdsharma26 thanks so much for taking a look! I will look into these comments soon

**Comment by rdsharma26** [comment]
@arsenalgunnershubert777 Let us know if you had a chance to look at the comments and if you can submit an update to the PR. If you do so, be sure to sync and rebase against the tracking branch to avoid conflicts.
